bridges/snowbridge/pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	#[pallet::storage]
+	pub type RewardsMapping<T: Config> =
+		StorageMap<_, Identity, AccountIdOf<T>, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::WeightInfo::claim(), DispatchClass::Operational))]
+		pub fn claim(
+			origin: OriginFor<T>,
+			deposit_address: AccountIdOf<T>,
+			value: BalanceOf<T>,
+			message_id: H256,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			Self::process_claim(account_id, deposit_address, value, message_id)?;
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn process_claim(
+			account_id: AccountIdOf<T>,
+			deposit_address: AccountIdOf<T>,
+			value: BalanceOf<T>,
+			message_id: H256,
+		) -> DispatchResult {
+			// Check if the claim value is equal to or less than the accumulated balance.
+			let reward_balance = RewardsMapping::<T>::get(account_id.clone());
+			if value > reward_balance {
+				return Err(Error::<T>::InsufficientFunds.into());
+			}
+
+			let reward_asset = snowbridge_core::location::convert_token_address(
+				T::EthereumNetwork::get(),
+				T::WethAddress::get(),
+			);
+			let cost2: u128 =
+				TryInto::<u128>::try_into(value).map_err(|_| Error::<T>::InvalidAmount)?;
+			let deposit: Asset = (reward_asset, cost2).into();
+			let beneficiary: Location =
+				Location::new(0, Parachain(T::AssetHubParaId::get().into()));
+			let bridge_location = Location::new(2, GlobalConsensus(T::EthereumNetwork::get()));
+
+			let xcm_fee: u128 = 10_000_000_000; // TODO not sure what this should be
+			let asset_hub_fee_asset: Asset = (Location::parent(), xcm_fee).into();
+
+			let fee: BalanceOf<T> = xcm_fee.try_into().map_err(|_| Error::<T>::InvalidFee)?;
+			burn_fees::<T::AssetTransactor, BalanceOf<T>>(T::AssetHubParaId::get().into(), fee)?;
+
+			let xcm: Xcm<()> = alloc::vec![
+				// Teleport required fees.
+				ReceiveTeleportedAsset(asset_hub_fee_asset.clone().into()),
+				// Pay for execution.
+				BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
+				DescendOrigin(PalletInstance(80).into()),
+				UniversalOrigin(GlobalConsensus(T::EthereumNetwork::get())),
+				ReserveAssetDeposited(deposit.clone().into()),
+				DepositAsset { assets: Definite(deposit.into()), beneficiary },
+				SetAppendix(Xcm(alloc::vec![
+					RefundSurplus,
+					DepositAsset { assets: AllCounted(1).into(), beneficiary: bridge_location },
+				])),
+				SetTopic(message_id.into()),
+			]
+			.into();
+
+			// Deduct the reward from the claimable balance
+			RewardsMapping::<T>::mutate(account_id.clone(), |current_value| {
+				*current_value = current_value.saturating_sub(value);
+			});
+
+			let dest = Location::new(1, [Parachain(T::AssetHubParaId::get().into())]);
+			let (_xcm_hash, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```

Comment:
```suggestion
			let (_xcm_hash, xcm_delivery_fee) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```
The cost of delivering the message to AH needs to taken from the account of the origin. If they don't have the funds to cover the fee, then the process should be aborted.

Reply:
@vgeddes if I take the returned delivery fee from `send_xcm`, the AH XCM instruction has already been sent and aborting the process would not make sense, right? So should I rather just check the relayer balance on DOT is above a sensible (configured) number?

Reply:
Can get delivery cost with 
```
let (_, fee) = T::XcmSender::validate(&mut Some(dest), &mut Some(xcm))
```

https://github.com/Snowfork/polkadot-sdk/pull/142 FYI.